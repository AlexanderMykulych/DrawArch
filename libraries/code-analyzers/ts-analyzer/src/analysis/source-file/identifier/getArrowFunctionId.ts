import { ArrowFunction } from "ts-morph";
import { EntityType, FunctionEntity, CodeInfoType } from "../../types";
import { getNodeInfo } from "./getNodeId";
import { getParentId } from "./getParentId";

export function getArrowFunctionId(arrowFn: ArrowFunction): FunctionEntity {
  const sourceFile = arrowFn.getSourceFile()
  return {
    id: getParentId(arrowFn),
    entityType: EntityType.Function,
    typeString: arrowFn.getType().getText(),
    type: CodeInfoType.Entity,
    filePath: sourceFile.getFilePath(),
    source: getNodeInfo(sourceFile) ?? '<unknown>'
  }
}