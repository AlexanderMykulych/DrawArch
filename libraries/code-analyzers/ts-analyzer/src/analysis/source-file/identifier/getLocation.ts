import type { Node } from "ts-morph";
import type { EntityLocation } from "../../types";
import { getLogLocation } from "./getLogLocation";

export function getLocation(node: Node): EntityLocation {
  const sourceFile = node.getSourceFile()
  const start = sourceFile.getLineAndColumnAtPos(node.getStart())
  const end = sourceFile.getLineAndColumnAtPos(node.getEnd())
  return {
    start: node.getStart(),
    end: node.getEnd(),
    lineStart: start.line,
    columnStart: start.column,
    lineEnd: end.line,
    columnEnd: end.column,
    kind: node.getKindName(),
    ...getLogLocation(node),
  }
}