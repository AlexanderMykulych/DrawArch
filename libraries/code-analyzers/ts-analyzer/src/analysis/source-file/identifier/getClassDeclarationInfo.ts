import type { ClassDeclaration, MethodDeclaration } from "ts-morph";
import { EntityType, FunctionEntity, CodeInfoType, BaseEntity } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getClassDeclarationInfo(node: ClassDeclaration): BaseEntity {
  const sourceFile = node.getSourceFile()
  return {
    id: `${getParentId(node)}/${node.getName()}`,
    entityType: EntityType.Class,
    type: CodeInfoType.Entity,
    filePath: sourceFile.getFilePath(),
    ...getLocation(node),
  }
}