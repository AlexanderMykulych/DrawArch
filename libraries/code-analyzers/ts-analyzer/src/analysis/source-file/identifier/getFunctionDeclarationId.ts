import type { FunctionDeclaration } from "ts-morph";
import { CodeInfoType, EntityType, FunctionEntity } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getFunctionDeclarationId(fnDec: FunctionDeclaration): FunctionEntity {
  const sourceFile = fnDec.getSourceFile()

  return {
    id: `${getParentId(fnDec)}/${fnDec.getName()}`,
    entityType: EntityType.Function,
    filePath: sourceFile.getFilePath(),
    type: CodeInfoType.Entity,
    typeString: fnDec.getType().getText(),
    source: sourceFile
      ? getParentsInfo(sourceFile)
      : [],
    ...getLocation(fnDec),
  }
}