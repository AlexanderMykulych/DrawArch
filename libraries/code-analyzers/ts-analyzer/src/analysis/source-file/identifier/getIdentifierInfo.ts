import { Identifier, Node, SyntaxKind } from "ts-morph";
import { CodeInfoType, Entity, EntityType, RelationType } from "../../types";
import { getRelationBeetwenNodes } from "../relations/getRelationBeetwenNodes";
import { getLocation } from "./getLocation";
import { getNodeInfo } from "./getNodeId";
import { getParentId, getParentEntity, getParentsInfo, getParentNode } from "./getParentId";
import type { NodeCodeInfos } from "./types";

export function getIdentifierInfo(node: Identifier): NodeCodeInfos {
  const parentNode = getParentNode(node)
  const declarationNode = getDeclarationNode(node)
  
  const isParentNode = parentNode === declarationNode

  if (isParentNode) {
    return getNodeInfo(parentNode)
  }

  return getIdentifierEntity(node)
}

function getIdentifierEntity(node: Identifier): NodeCodeInfos {
  const implementations = node.getImplementations()

  let codeInfos: Entity[] = []
  if (implementations.length > 0) {
    codeInfos = implementations
      .map(x => x.getNode())
      .filter(x => x !== node)
      .flatMap(x => getNodeInfo(x)[0])
  }

  const parentEntity = getParentEntity(node)

  if (codeInfos.length > 0) {

    // @ts-ignore
    return [
      ...codeInfos,
      parentEntity,
      ...codeInfos
        .map(codeInfo => getRelationBeetwenNodes({
            from: parentEntity,
            to: codeInfo,
            type: RelationType.Use,
          })
        )
    ]
  }

  const declarationNode = node.getSymbol()?.getValueDeclaration() ?? node

  if (!declarationNode.isKind(SyntaxKind.Identifier)) {
    const declarationEntity = getNodeInfo(declarationNode)[0]

    const relations = codeInfos.map(x =>
      getRelationBeetwenNodes({
        to: x,
        from: declarationEntity,
        type: RelationType.Use,
        location: getLocation(declarationNode),
      })
    )

    return [
      declarationEntity,
      ...codeInfos,
      ...relations,
      getRelationBeetwenNodes({
        from: parentEntity,
        to: declarationEntity,
        type: RelationType.Parent,
        location: getLocation(node),
      }),
    ]
  }

  const location = getLocation(node)

  const identifierEntity: Entity = {
    type: CodeInfoType.Entity,
    entityType: EntityType.Identifier,
    // source: getParentsInfo(node) ?? [],
    id: `${getParentId(node)}/${node.getText()}`,
    filePath: node.getSourceFile().getFilePath(),
    ...location,
  }

  const relations = codeInfos.map(x =>
      getRelationBeetwenNodes({
        to: x,
        from: identifierEntity,
        type: RelationType.Use,
        location,
      })
    )

  return [
    identifierEntity,
    ...codeInfos,
    ...relations,
  ]
}

function getDeclarationNode(node: Identifier): Node | undefined {
  return node.getSymbol()?.getValueDeclaration()
}