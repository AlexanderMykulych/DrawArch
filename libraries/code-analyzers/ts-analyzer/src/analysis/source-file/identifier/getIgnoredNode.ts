import type { Node } from "ts-morph";
import { CodeInfoType, EntityType, UnknownEntity } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getIgnoredNode(node: Node): UnknownEntity {
  const sourceFile = node.getSourceFile()
  return {
    id: `<unknown>`,
    entityType: EntityType.Unknown,
    filePath: sourceFile.getFilePath(),
    type: CodeInfoType.Entity,
    code: node.getFullText(),
    meta_kind: node.getKindName(),
    source: sourceFile
      ? getParentsInfo(sourceFile)
      : [],
    ...getLocation(node),
  }
}