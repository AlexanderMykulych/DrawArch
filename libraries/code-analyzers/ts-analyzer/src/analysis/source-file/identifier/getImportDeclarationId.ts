import type { ImportDeclaration } from "ts-morph";
import { BaseEntity, CodeInfoType, EntityType } from "../../types";
import { getLocation } from "./getLocation";
import { getParentsInfo } from "./getParentId";

export function getImportDeclarationId(importDec: ImportDeclaration): BaseEntity {
  const sourceFile = importDec?.getModuleSpecifierSourceFile();
  const id = importDec.getModuleSpecifierSourceFile()?.getFilePath()?.toString() ?? importDec.getText()
  return {
    id,
    entityType: EntityType.ImportDeclaration,
    filePath: importDec.getSourceFile()?.getFilePath() ?? '<unknow>',
    type: CodeInfoType.Entity,
    source: sourceFile
      ? getParentsInfo(sourceFile)
      : [],
    ...getLocation(importDec),
  }
}