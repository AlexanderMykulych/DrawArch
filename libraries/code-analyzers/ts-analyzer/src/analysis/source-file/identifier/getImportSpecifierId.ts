import type { ImportSpecifier } from "ts-morph";
import { BaseEntity, CodeInfoType, EntityType } from "../../types";
import { getLocation } from "./getLocation";
import { getParentId, getParentsInfo } from "./getParentId";

export function getImportSpecifierId(node: ImportSpecifier): BaseEntity {
  const moduleSourceFile = node.getSourceFile()
  return {
    id: `${getParentId(node)}/${node.getName()}`,
    entityType: EntityType.Variable,
    filePath: moduleSourceFile?.getFilePath() ?? '<unknown>',
    type: CodeInfoType.Entity,
    // source: moduleSourceFile
    //   ? getParentsInfo(moduleSourceFile)
    //   : [],
    ...getLocation(node),
  }
}