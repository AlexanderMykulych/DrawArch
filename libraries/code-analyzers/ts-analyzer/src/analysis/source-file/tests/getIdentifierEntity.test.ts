import { Project, SyntaxKind } from "ts-morph";
import { expect, test } from "vitest";
import { getIdentifierInfo } from "../identifier/getIdentifierInfo";

test('should', () => {
  const project = new Project({
    useInMemoryFileSystem: true,
  });
  const mainFile = project.createSourceFile('main.ts', 'import { sum } from "./utils"')
  project.createSourceFile('utils.ts', 'export function sum(a: number, b: number) { return a + b }"')

  const identifier = mainFile.getDescendantsOfKind(SyntaxKind.Identifier).find(x => x.getText() === 'sum')
  
  const res = identifier ? getIdentifierInfo(identifier) : null

  expect(res).toMatchSnapshot()
})
