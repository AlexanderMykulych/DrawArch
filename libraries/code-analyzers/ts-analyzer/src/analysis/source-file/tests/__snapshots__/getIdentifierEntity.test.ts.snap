// Vitest Snapshot v1

exports[`should 1`] = `
[
  {
    "code": " { return a + b }",
    "columnEnd": 59,
    "columnStart": 1,
    "end": 58,
    "entityType": "function",
    "filePath": "/utils.ts",
    "id": "/utils.ts/sum",
    "kind": "FunctionDeclaration",
    "lineEnd": 1,
    "lineStart": 1,
    "logEnd": 58,
    "logStart": 42,
    "start": 0,
    "type": "entity",
    "typeString": "typeof import(\\"/utils\\").sum",
  },
  {
    "entityType": "file",
    "filePath": "/main.ts",
    "id": "/main.ts",
    "type": "entity",
  },
  {
    "from": "/main.ts",
    "id": "/main.ts->/utils.ts/sum",
    "relationType": "use",
    "to": "/utils.ts/sum",
    "type": "relationship",
  },
]
`;
